mlir/include/mlir/Interfaces/SideEffectInterfaces.h

```diff
@@ -149,11 +149,20 @@ class EffectInstance {
                  Resource *resource = DefaultResource::get())
       : effect(effect), resource(resource), value(value), stage(0),
         effectOnFullRegion(false) {}
+  EffectInstance(EffectT *effect, OpOperand *opd,
```

Comment:
The list of constructors gets longer and longer... Do you see a way to shorten it a bit?

Reply:
As previously discussed, I have added OpResult and BlockArgument, and reimplemented it using templates. Please review again